File: java/src/irrgarten/UI/TextUI.java

package irrgarten.UI;

import irrgarten.Directions;
import irrgarten.GameState;
import java.util.Scanner;


public class TextUI {
    
    private static Scanner in = new Scanner(System.in);
    
    private char readChar() {
        String s = in.nextLine();
        return s.charAt(0);
    }
    

    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }    
        return direction;
    }
    
    public void showGame(GameState gameState) {  

        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println("========== IRRGARTEN ==========");
        System.out.println("¿Ganador?: " + gameState.getWinner());
        System.out.println("Jugador actual: " + gameState.getCurrentPlayer());
        System.out.println();
        System.out.println("------ Laberinto ------");
        System.out.println(gameState.getLabyrinth().toString());
        System.out.println();
        System.out.println("------ Jugadores ------");
        System.out.println(gameState.getPlayers().toString());
        System.out.println();
        System.out.println("------ Monstruos ------");
        System.out.println(gameState.getMonsters().toString());
        System.out.println();
        System.out.println("------ Registro ------");
        System.out.println(gameState.getLog());
        System.out.println("===============================");
    }
    
}

File: ruby/UI/textUI.rb
#encoding: utf-8

require 'io/console'
require_relative '../Directions'
require_relative '../Player'

module UI

  class TextUI


    #https://gist.github.com/acook/4190379
    def read_char
      STDIN.echo = false
      STDIN.raw!
    
      input = STDIN.getc.chr
      if input == "\e" 
        input << STDIN.read_nonblock(3) rescue nil
        input << STDIN.read_nonblock(2) rescue nil
      end
    ensure
      STDIN.echo = true
      STDIN.cooked!
    
      return input
    end

    

    def next_move
      print "Where? "
      got_input = false
      while (!got_input)
        c = read_char
        case c
          when "\e[A"
            puts "UP ARROW"
            output = Irrgarten::Directions::UP
            got_input = true
          when "\e[B"
            puts "DOWN ARROW"
            output = Irrgarten::Directions::DOWN
            got_input = true
          when "\e[C"
            puts "RIGHT ARROW"
            output = Irrgarten::Directions::RIGHT
            got_input = true
          when "\e[D"
            puts "LEFT ARROW"
            output = Irrgarten::Directions::LEFT
            got_input = true
          when "\u0003"
            puts "CONTROL-C"
            got_input = true
            exit(1)
          else
            #Error
        end
      end
      output
    end
   

    # Método para mostrar el estado del juego
    def show_game(game_state)
      #system("clear") || system("cls")  # Limpia la pantalla (UNIX o Windows)
      puts
      puts "========== IRRGARTEN =========="
      puts "¿Ganador?: #{game_state.winner}"
      puts "Jugador actual: #{game_state.current_player}"
      puts
    
      puts "------ Laberinto ------"
      puts game_state.labyrinth.to_s  # Asumimos que imprime el tablero o mapa
    
      puts
      puts "------ Jugadores ------"
      puts game_state.players.to_s  # Asumimos que ya es string formateado


      puts
      puts "------ Monstruos ------"
      puts game_state.monsters.to_s  # Asumimos que ya es string formateado
      
    
      puts
      puts "------ Registro ------"
      puts game_state.log

      if game_state.winner
        puts "\n¡Ganador de la partida: Player " + game_state.current_player.to_s + "!"
      end
    
      puts "==============================="
      puts 

    end # show_game

  end # class TextUI


end # module   


